Draft: proc macro design doc (#1049)
software-mansion/scarb: design/01-proc-macro.md

# Procedural Macros

## Summary

> We are actively thinking how to allow using user-developed Cairo compiler plugins ([`MacroPlugin`]) in Scarb-powered projects. This post describes our current idea on how to approach this and outlines implementation steps. We would like to invite the community to share feedback!

Reviewer: `MacroPlugin` is a missing link

This document proposes an evolution of Cairo’s compiler plugins system. The primary motivation behind this proposal is to allow dynamic loading of compiler plugins per compilation unit. This document specifies syntax for the definition of procedural macros, a high-level view of their implementation in the compiler, and outlines how they interact with the compilation process.

## Context

### Rust procedural macros

This post assumes reader knowledge on Rust's procedural macros and how they are implemented. A good walkthrough has been published some time ago on IntelliJ Rust blog:

1. https://blog.jetbrains.com/rust/2022/03/18/procedural-macros-under-the-hood-part-i/

> 2. https://blog.jetbrains.com/rust/2022/07/07/procedural-macros-under-the-hood-part-ii/

Author: Looking through rust release notes etc I have not found any information about changes that I would find important. 

I know there is some pre-rfc work on compiling proc-macros to wasm though, see https://internals.rust-lang.org/t/pre-rfc-sandboxed-deterministic-reproducible-efficient-wasm-compilation-of-proc-macros/19359

TBH I feel like even if this articles would in fact become outdated, it would still be just enough of a context for this doc reader. WDYT?

> 2. https://blog.jetbrains.com/rust/2022/07/07/procedural-macros-under-the-hood-part-ii/

Reviewer: If nothing drastic changed, then these articles are relevant. Great!

Scarb currently simply assumes a fixed, predefined plugin sets for `lib`, `starknet-contract` and `test` targets. This interface is not stable and is subject to changes in future compiler versions.

### Plugins API in combination with Scarb

Because of the use of shared reference to [`SyntaxGroup`], this API is hostile to attempts to externalize plugins into separate binaries/processes that communicate via some kind of RPC. The fact that AST structure is not stable does not help either. On the other hand, access to [`SyntaxGroup`] allows for powerful operations and is highly desired.

The only feasible way of loading plugins is to load them directly into Scarb compiler's memory. This brings another problem though: Rust does not have a stable ABI, and thus such system would have to ensure that the plugin library was compiled with **exactly** the same Rust toolchain, for example `1.68.1-x86_64-pc-windows-msvc`.

### Declaring package dependencies on plugins

Scarb allows packages to declare dependency on a plugin through the same mechanism as on standard Cairo libraries. The plugin must be defined as a Scarb package defining `cairo-plugin` target in the package manifest.

> Since currently there is no mechanism to supply plugins implementation, it must be compiled into Scarb itself. Only `starknet` plugin is supplied this way.

Reviewer: Test plugin also is

## Proposal

This document proposes an introduction of new compiler plugins type, called procedural macros, as described below.

> ### Plugins will be compiled into shared libraries

Reviewer: Let's also try to make an alternative PoC which uses Extism. It will probably have a slightly worse performance; but we could for example prebuild some trusted plugins on the registry side which could be a big win.

And also, Extism will not require Rust version parity.

> - Plugins will be implemented in Rust.

Reviewer: ```suggestion
- Procedural macros will be implemented in Rust.
```

Reviewer: forgot to say that this applies to all instances later on in this document 😅

- Procedural macros will be implemented in Rust.

> - Procedural macros will be compiled into shared libraries, that Scarb can load at the runtime through dynamic loading mechanism of the host ope rating system.

Reviewer: ```suggestion
- Procedural macros will be compiled into shared libraries, that Scarb can load at the runtime through dynamic loading mechanism of the host operating system.
```

- Scarb will use [`libloading` crate](https://docs.rs/libloading/latest/libloading/), which exposes a cross-platform interface for interactions with platform dynamic library loading primitives, to load shared libraries into memory.

### Plugins will be special Scarb packages containing Rust code

- A plugin will be a special Scarb package, which will use a new native target type: `cairo-plugin`.
- The plugin package will contain Rust plugin source code, which will be shipped to users on Scarb project resolution.
- The plugin source code will be compiled on Scarb users system only.
- Enabling this target will mean that this package does not contain any Cairo code.
- This target will be *exclusive*:
    - It will block defining other targets for the package, not even `lib`.
    - It will also not be possible to declare dependencies, or specify Cairo compiler settings, it won't make sense for these packages.

> - Scarb will automatically generate a `Cargo.toml` manifest file, based on contents of `Scarb.toml` manifest. The Cargo manifest will be used for the plugin compilation.

Reviewer: This does not tell _when_ `Cargo.toml` will be generated. I think it should be part of workspace resolution (i.e. `scarb fetch`). Furthermore, I believe calling `cargo fetch` for all plugins in the workspace should also be done during this phase.

- Scarb will automatically generate a `Cargo.toml` manifest file, based on contents of `Scarb.toml` manifest. The Cargo manifest will be used for the procedural macro compilation.
    - The `lib.crate-type` will be set `dylib`
- The Cargo manifest generation, followed by the Cargo resolution of procedural macro packages, will be handled during the Scarb workspace resolution. 
- The compiled procedural macros will be cached on the user filesystem in Scarb cache directory, to ensure no recompilation on each Scarb project build is required.

**Example:**

Scarb manifest file:

```toml
[package]
name = "ignore-plugin"
version = "0.1.0"

[cairo-plugin]
```

Automatically generated Cargo manifest file:

```toml

> [package]
> name = "ignore-plugin"
> version = "0.1.0"
> 
> [lib]
> crate-type = "dylib"
> 
> [dependencies]
> cairo-lang-macro = "1.0.0-alpha.7"

Reviewer: ```suggestion
# Code generated by scarb fetch; DO NOT EDIT.
#
# The Cargo.toml file for Cairo procedural macros crates is entirely managed by Scarb,
# the Cairo package manager. Scarb regenerates this file whenever it processes Scarb.toml.
# The simplest way to trigger this, is to run the `scarb fetch` command.
#
# If you want to provide custom data in this file, include it in Scarb.toml under
# [tool.cargo.pkg] section. For example, to add Rust dependencies, type the following:
#
#     [tool.cargo.pkg.dependencies]
#     serde = "1"
#     serde_json = "1"

[package]
name = "ignore-plugin"
version = "0.1.0"

[lib]
crate-type = "dylib"

[dependencies]
cairo-lang-macro = "1.0.0-alpha.7"
```
